src/msg/simple/SimpleMessenger.cc

```diff
@@ -396,19 +396,23 @@ bool SimpleMessenger::verify_authorizer(Connection *con, int peer_type,
 
 ConnectionRef SimpleMessenger::get_connection(const entity_inst_t& dest)
 {
-  Mutex::Locker l(lock);
-  if (my_inst.addr == dest.addr) {
-    // local
-    return local_connection;
+  {
+    Mutex::Locker l(lock);
+    if (my_inst.addr == dest.addr) {
+      // local
+      return local_connection;
+    }
   }
-
   // remote
   while (true) {
     Pipe *pipe = _lookup_pipe(dest.addr);
     if (pipe) {
       ldout(cct, 10) << "get_connection " << dest << " existing " << pipe << dendl;
     } else {
-      pipe = connect_rank(dest.addr, dest.name.type(), NULL, NULL);
+      Mutex::Locker l(lock);
+      pipe = _lookup_pipe(dest.addr);
```

Comment:
I guess this is wrong as unordered_map is not thread safe...

So in L407 , when you _lookup_pipe without holding the lock, some other thread may creating a Pipe, which will break the unordered_map.

Given that, could you provide some explanation why this could help?